src/neuralnet/neuron_layer/gru.cc

```diff
@@ -0,0 +1,265 @@
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*
+*************************************************************/
+
+#include "singa/neuralnet/neuron_layer/gru.h"
+
+#include <glog/logging.h>
+#include "singa/utils/singleton.h"
+#include "singa/utils/math_blob.h"
+#include "singa/utils/singa_op.h"
+
+#include <iostream>
+using namespace std;
+
+namespace singa {
+
+using std::vector;
+
+GRULayer::~GRULayer() {
+  delete weight_z_hx_;
+  delete weight_z_hh_;
+  delete bias_z_; 
+
+  delete weight_r_hx_;
+  delete weight_r_hh_;
+  delete bias_r_;
+
+  delete weight_c_hx_; 
+  delete weight_c_hh_;
+  delete bias_c_; 
+
+  delete update_gate;
+  delete reset_gate;
+  delete new_memory;
+}
+
+void GRULayer::Setup(const LayerProto& conf,
+    const vector<Layer*>& srclayers) {
+  Layer::Setup(conf, srclayers);
+  CHECK_LE(srclayers.size(), 2);
+  const auto& src = srclayers[0]->data(this);
+  
+  batchsize_ = src.shape()[0]; // size of batch
+  vdim_ = src.count() / (batchsize_); // dimension of input
+
+  hdim_ = layer_conf_.gru_conf().dim_hidden(); // dimension of hidden state
+
+  data_.Reshape(vector<int>{batchsize_, hdim_});
+  grad_.ReshapeLike(data_);
+
+  // Initialize the parameters
+  weight_z_hx_ = Param::Create(conf.param(0));
+  weight_z_hh_ = Param::Create(conf.param(1));
+  bias_z_ = Param::Create(conf.param(2));
+
+  weight_r_hx_ = Param::Create(conf.param(3));
+  weight_r_hh_ = Param::Create(conf.param(4));
+  bias_r_ = Param::Create(conf.param(5)); 
+
+  weight_c_hx_ = Param::Create(conf.param(6));
+  weight_c_hh_ = Param::Create(conf.param(7));
+  bias_c_ = Param::Create(conf.param(8));
+  
+  
+  weight_z_hx_->Setup(vector<int>{hdim_, vdim_});
+  weight_z_hh_->Setup(vector<int>{hdim_, hdim_});
+  bias_z_->Setup(vector<int>{hdim_});
+
+  weight_r_hx_->Setup(vector<int>{hdim_, vdim_});
+  weight_r_hh_->Setup(vector<int>{hdim_, hdim_});
+  bias_r_->Setup(vector<int>{hdim_});
+
+  weight_c_hx_->Setup(vector<int>{hdim_, vdim_});
+  weight_c_hh_->Setup(vector<int>{hdim_, hdim_});
+  bias_c_->Setup(vector<int>{hdim_});
+
+  update_gate = new Blob<float>(batchsize_, hdim_);
+  reset_gate = new Blob<float>(batchsize_, hdim_);
+  new_memory = new Blob<float>(batchsize_, hdim_);
+
+}
+
+void GRULayer::ComputeFeature(int flag,
+    const vector<Layer*>& srclayers) {
+	CHECK_LE(srclayers.size(), 2);
+
+	// Do transpose
+	Blob<float> *w_z_hx_t = Transpose (weight_z_hx_->data());
+	Blob<float> *w_z_hh_t = Transpose (weight_z_hh_->data());
+	Blob<float> *w_r_hx_t = Transpose (weight_r_hx_->data());
+	Blob<float> *w_r_hh_t = Transpose (weight_r_hh_->data());
+	Blob<float> *w_c_hx_t = Transpose (weight_c_hx_->data());
+	Blob<float> *w_c_hh_t = Transpose (weight_c_hh_->data());
+
```

Comment:
By the way, you mentioned "there will be". When this new function will be added?